operators/src/concurrency/thread_pool.rs

```diff
@@ -0,0 +1,339 @@
+        (self.task_fn)()
+    }
+}
+
+/// A worker thread pool for compute-heavy tasks
+///
+/// TODO: increase threads to a certain maximum number if current threads don't produce results fast enough?
+/// TODO: schedule tasks based on group context?
+///
+#[derive(Debug)]
+pub struct ThreadPool {
+    degree_of_parallelism: usize,
+    global_queue: Arc<Injector<Task>>,
+    stealers: Vec<Stealer<Task>>,
+    threads: Vec<JoinHandle<()>>,
+    next_group_id: AtomicCell<usize>,
+}
+
+impl ThreadPool {
+    pub fn new(number_of_threads: usize) -> Self {
+        assert!(
+            number_of_threads > 0,
+            "There must be at least one thread for the thread pool"
+        );
+
+        let worker_deques: Vec<Worker<Task>> =
+            (0..number_of_threads).map(|_| Worker::new_fifo()).collect();
+
+        let mut thread_pool = Self {
+            degree_of_parallelism: number_of_threads,
+            global_queue: Arc::new(Injector::new()),
+            stealers: worker_deques.iter().map(Worker::stealer).collect(),
+            threads: Vec::with_capacity(number_of_threads),
+            next_group_id: AtomicCell::new(0),
+        };
+
+        for worker_deque in worker_deques {
+            let global_queue = thread_pool.global_queue.clone();
+            let stealers = thread_pool.stealers.clone();
+
+            thread_pool.threads.push(std::thread::spawn(move || {
+                Self::await_work(&worker_deque, &global_queue, &stealers);
+            }))
+        }
+
+        thread_pool
+    }
+
+    fn await_work(local: &Worker<Task>, global: &Injector<Task>, stealers: &[Stealer<Task>]) {
+        let backoff = Backoff::new();
+
+        loop {
+            // Pop a task from the local queue, if not empty.
+            let task = local.pop().or_else(|| {
+                // Otherwise, we need to look for a task elsewhere.
+                iter::repeat_with(|| {
+                    // Try stealing a batch of tasks from the global queue.
+                    global
+                        .steal_batch_and_pop(local)
+                        // Or try stealing a task from one of the other threads.
+                        .or_else(|| stealers.iter().map(Stealer::steal).collect())
+                })
+                // Loop while no task was stolen and any steal operation needs to be retried.
+                .find(|s| !s.is_retry())
+                // Extract the stolen task, if there is one.
+                .and_then(Steal::success)
+            });
+
+            if let Some(task) = task {
+                // TODO: recover panics
+                task.call_once();
+                backoff.reset();
+            } else {
+                // TODO: sleep instead of busy waiting?
+                backoff.snooze();
+            }
+        }
+    }
+
+    pub fn create_context(&self) -> ThreadPoolContext {
+        ThreadPoolContext::new(self, self.next_group_id.fetch_add(1))
+    }
+
+    fn compute(&self, task: Task) {
```

Comment:
what about tasks that return something? Would be nice to get a future one can await on.

Reply:
Yeah, that would be somewhat more complicated but I guess also possible. Would compute then return some handle to the yet-to-be-computed result? What would this handle be? Is this a common operation, where we spawn tasks that return something?

Reply:
Aren't we always interested in the result of a computation? ^^ In the tests you always pass a shared atomic reference for returning the result. Is this always feasible? E.g. when computing some geometry operation like buffer for a feature collection and each feature spawns a task.

Reply:
ce760c0 

I'm not quite sure that `Future`s are the appropriate return type here since it mixed async and normal computations, but it works.